File: insertion_sort.c
#include "stdio.h"
#include "stdlib.h"
#include "stdbool.h"

bool sort(int **a, int size);
bool swap(int *a, int *b);

int main()
{
	int *a = calloc(9, sizeof(int));
	a[0] = 2;
	a[1] = 1; 
	a[2] = 5;
	a[3] = 4;
	a[4] = 9;
	a[5] = 8;
	a[6] = 6;
	a[7] = 7;
	a[8] = 3;
	sort(&a, 9);

	for (int i = 0; i < 9; i++)
		printf("%d ", a[i]);

	printf("\n");

	free(a);
	return 0;
}

bool sort(int **a, int size)
{
	if (a == NULL || (*a) == NULL)
		return false;

	for (int i = 0; i < size; i++)
	{
		int j = i;
		while (j > 0)
			if ((*a)[j] < (*a)[j - 1])
			{
				swap(&((*a)[j]), &((*a)[j - 1]));
				j--;
			} 
			else 
				break;
	}
}

bool swap(int *a, int *b)
{
	int t = *a;
	*a = *b;
	*b = t;
}

File: InsertionSort.java
public class InsertionSort {

    private int shifts;

    public int[] swap(int[] array, int i, int j) {
        int t = array[i];
        array[i] = array[j];
        array[j] = t;
        return array;
    }

    public int[] shift(int[] array, int i) {
        if (i == 0)
            return array;

        while (i > 0) {
            if (array[i] < array[i - 1]) {
                array = swap(array, i, i - 1);
                i--;
                shifts++;
            } else
                break;
        }

        return array;
    }

    public int[] sort(int[] array, int i) {
        if (i == 0)
            return array;

        sort(array, i - 1);
        shift(array, i - 1);

        return array;
    }

    public static void main(String[] args) {
        //int[] array = new int[] {4, 3, 2, 1};
        //int[] array = new int[] {1, 1, 1, 2, 2};
        int[] array = new int[] {2, 1, 3, 1, 2};

        InsertionSort is = new InsertionSort();
        array = is.sort(array, array.length);

        for (int i = 0; i < array.length; i++) {
            System.out.print(array[i] + " ");
        }

        System.out.println();

        System.out.println("Shifts: " + is.shifts);
    }
}
